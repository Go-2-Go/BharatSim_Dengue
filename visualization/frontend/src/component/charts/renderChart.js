import chartConfig from '../../config/chartConfig';

function renderChart(chartType, chartProps) {
  return chartConfig[chartType].chart(chartProps);
}

export default renderChart;
