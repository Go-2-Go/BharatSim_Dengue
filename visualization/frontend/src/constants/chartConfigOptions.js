import React from 'react';

import XAxisChartConfig from '../component/chartConfigOptions/XAxisChartConfig';
import YAxisChartConfig from '../component/chartConfigOptions/YAxisChartConfig';

export const chartConfigOptions = {
  X_AXIS: 'xAxis',
  Y_AXIS: 'yAxis',
};

export function getChartConfigComponent(
  chartConfigType,
  { headers, updateConfigState, errors, values },
) {
  return {
    [chartConfigOptions.X_AXIS]: (
      <XAxisChartConfig
        headers={headers}
        updateConfigState={updateConfigState}
        configKey={chartConfigOptions.X_AXIS}
        error={errors[chartConfigOptions.X_AXIS]}
        value={values[chartConfigOptions.X_AXIS]}
      />
    ),
    [chartConfigOptions.Y_AXIS]: (
      <YAxisChartConfig
        headers={headers}
        updateConfigState={updateConfigState}
        configKey={chartConfigOptions.Y_AXIS}
        error={errors[chartConfigOptions.Y_AXIS]}
        value={values[chartConfigOptions.Y_AXIS]}
      />
    ),
  }[chartConfigType];
}
