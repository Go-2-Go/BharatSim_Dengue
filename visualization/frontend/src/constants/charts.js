export const chartTypes = {
  LINE_CHART: 'lineChart',
  BAR_CHART: 'barChart',
};
