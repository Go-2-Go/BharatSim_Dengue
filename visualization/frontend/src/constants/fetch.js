export const httpMethods = {
  GET: 'get',
  POST: 'post',
};

export const contentTypes = {
  FILE: 'multipart/form-data',
};
