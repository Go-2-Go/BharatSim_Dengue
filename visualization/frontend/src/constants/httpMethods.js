export const httpMethods = {
  GET: 'get',
  POST: 'post',
};
