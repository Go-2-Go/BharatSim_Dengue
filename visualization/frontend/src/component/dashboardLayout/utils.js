import React from 'react';
import renderChart from '../charts/renderChart';

export function renderElement(element) {
  return (
    <div key={element.i} data-grid={element} data-testid={element.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(element.chartType, { config: element.config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 2) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 2,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
