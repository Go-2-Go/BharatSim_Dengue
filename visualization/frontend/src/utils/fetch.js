import axios from 'axios';
import { httpMethods } from '../constants/fetch';

let shouldShowError;

function commonErrorHandler(err) {
  shouldShowError(true);
}

function initApiConfig({ setError }) {
  shouldShowError = setError;
}

function fetchData({ url, method = httpMethods.GET, headers, data, query, customErrorHandler }) {
  return axios({ url, method, headers, data, params: query })
    .then((res) => res.data)
    .catch((err) => {
      if (customErrorHandler) {
        customErrorHandler(err);
        return;
      }
      commonErrorHandler(err);
    });
}

function uploadData({ url, headers, data, query, customErrorHandler }) {
  return fetchData({
    url,
    method: httpMethods.POST,
    headers,
    data,
    params: query,
    customErrorHandler,
  });
}

export { fetchData, uploadData, initApiConfig };
