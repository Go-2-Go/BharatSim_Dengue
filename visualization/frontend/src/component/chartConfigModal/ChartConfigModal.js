import React from 'react';
import PropTypes from 'prop-types';

import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Typography,
  withStyles,
} from '@material-ui/core';
import Dropdown from '../../uiComponent/Dropdown';
import renderChartConfig from '../chartConfigOptions/renderChartConfig';
import styles from './chartConfigModalCss';

import useFetch from '../../hook/useFetch';
import useForm from '../../hook/useForm';
import { fetch } from '../../utils/fetch';
import { convertObjectArrayToOptions } from '../../utils/helper';
import chartConfigs from '../../config/chartConfigs';
import { url } from '../../utils/url';
import { datasourceValidator, getChartConfigValidator } from '../../utils/validators';

function createValidatorsSchema(configOptions) {
  const chartConfigValidatorSchema = {};
  configOptions.forEach((chartConfig) => {
    chartConfigValidatorSchema[chartConfig] = getChartConfigValidator(chartConfig);
  });
  return chartConfigValidatorSchema;
}

function ChartConfigModal({ open, onCancel, onOk, chartType, classes }) {
  const [headers, setHeaders] = React.useState([]);
  const datasources = useFetch({ url: url.DATA_SOURCES });
  const { validateAndSetValue, errors, shouldEnableSubmit, onSubmit } = useForm({
    ...createValidatorsSchema(chartConfigs[chartType].configOptions),
    dataSource: datasourceValidator,
  });

  if (!datasources) {
    return null;
  }

  const handleDataSourceChange = async (value) => {
    validateAndSetValue('dataSource', value);
    const csvHeaders = await fetch({ url: url.getHeaderUrl(value) });
    setHeaders(csvHeaders.headers);
  };

  const handleOk = () => {
    onSubmit((value) => onOk(value));
  };

  const chartConfigProps = { headers, updateConfigState: validateAndSetValue, errors };

  return (
    <Dialog open={open} onClose={onCancel} aria-labelledby="form-dialog-title">
      <DialogTitle id="form-dialog-title">Chart Config</DialogTitle>

      {datasources.dataSources.length === 0 ? (
        <Box p={10}>
          <Typography>No data source present, upload data source</Typography>
        </Box>
      ) : (
        <>
          <DialogContent>
            <Box className={classes.root}>
              <Dropdown
                options={convertObjectArrayToOptions(datasources.dataSources, '_id', 'name')}
                onChange={handleDataSourceChange}
                id="dropdown-dataSources"
                label="select data source"
                error={errors.dataSource || ''}
              />
              {!!headers.length &&
                renderChartConfig(chartConfigs[chartType].configOptions, chartConfigProps)}
            </Box>
          </DialogContent>
          <DialogActions>
            <Button onClick={onCancel} variant="contained" color="secondary">
              Cancel
            </Button>
            <Button
              onClick={handleOk}
              variant="contained"
              color="primary"
              data-testid="button-ok"
              disabled={!shouldEnableSubmit()}
            >
              Ok
            </Button>
          </DialogActions>
        </>
      )}
    </Dialog>
  );
}

ChartConfigModal.propTypes = {
  open: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  onOk: PropTypes.func.isRequired,
  chartType: PropTypes.string.isRequired,
  classes: PropTypes.shape({
    root: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles)(ChartConfigModal);
export { ChartConfigModal };
