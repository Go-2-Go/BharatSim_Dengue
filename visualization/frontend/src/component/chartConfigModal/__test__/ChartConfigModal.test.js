import React from 'react';

import { fireEvent, render, waitFor, within } from '@testing-library/react';

import ChartConfigModal from '../ChartConfigModal';
import { fetch } from '../../../utils/fetch';
import { chartTypes } from '../../../constants/charts';
import { selectDropDownOption } from '../../../testUtil';

jest.mock('../../../utils/url', () => ({
  url: {
    getHeaderUrl: jest.fn().mockReturnValue('/headers'),
    DATA_SOURCES: '/datasources',
  },
}));

jest.mock('../../../utils/fetch', () => ({
  fetch: jest.fn(({ url }) => {
    if (url === '/headers') {
      return Promise.resolve({ headers: ['x-header', 'y-header'] });
    }
    if (url === '/datasources') {
      return Promise.resolve({
        dataSources: [
          { _id: 'id1', name: 'modelone' },
          { _id: 'id2', name: 'modeltwo' },
        ],
      });
    }
    return Promise.reject();
  }),
}));

describe('<ChartConfigModal />', () => {
  const props = {
    onCancel: jest.fn(),
    onOk: jest.fn(),
    open: true,
    chartType: chartTypes.LINE_CHART,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match a snapshot for <ChartConfigModal />', async () => {
    render(<ChartConfigModal {...props} />);

    await waitFor(() => document.querySelector('.MuiPaper-root'));

    expect(document.querySelector('.MuiPaper-root')).toMatchSnapshot();
  });

  it('should be show x-axis and y-axis dropdown if data source selected', async () => {
    render(<ChartConfigModal {...props} />);

    await waitFor(() => document.querySelector('.MuiPaper-root'));

    const configModal = within(document.querySelector('.MuiPaper-root'));

    selectDropDownOption(configModal, 'dropdown-dataSources', 'id1');

    await waitFor(() => configModal.queryByText('dropdown-x'));

    const xDropdown = within(configModal.getByTestId('dropdown-x'));
    const yDropdown = within(configModal.getByTestId('dropdown-y'));

    await waitFor(() => {
      expect(xDropdown.getByRole('button')).toBeInTheDocument();
      expect(yDropdown.getByRole('button')).toBeInTheDocument();
    });
  });

  it('should be hide x-axis and y-axis dropdown if data source is not selected', async () => {
    render(<ChartConfigModal {...props} />);

    await waitFor(() => document.querySelector('.MuiPaper-root'));

    const configModal = within(document.querySelector('.MuiPaper-root'));

    const xDropdown = configModal.queryByTestId('dropdown-x');
    const yDropdown = configModal.queryByTestId('dropdown-y');

    expect(xDropdown).toBeNull();
    expect(yDropdown).toBeNull();
  });

  it('should pass selected chart configs to onOK callback', async () => {
    render(<ChartConfigModal {...props} />);

    await waitFor(() => document.querySelector('.MuiPaper-root'));

    const configModal = within(document.querySelector('.MuiPaper-root'));

    selectDropDownOption(configModal, 'dropdown-dataSources', 'id1');

    await waitFor(() => expect(fetch).toBeCalledTimes(2));

    selectDropDownOption(configModal, 'dropdown-x', 'x-header');

    selectDropDownOption(configModal, 'dropdown-y', 'y-header');

    const okButton = configModal.getByText(/Ok/i);

    fireEvent.click(okButton);

    expect(props.onOk).toHaveBeenCalledWith({
      dataSource: 'id1',
      xAxis: 'x-header',
      yAxis: 'y-header',
    });
  });

  it('should close select config modal on cancel click', async () => {
    render(<ChartConfigModal {...props} />);

    await waitFor(() => document.querySelector('.MuiPaper-root'));

    const configModal = within(document.querySelector('.MuiPaper-root'));

    const cancelButton = configModal.getByText(/Cancel/i);
    fireEvent.click(cancelButton);

    expect(props.onCancel).toHaveBeenCalled();
  });

  it('should provide empty element if csv header is null', async () => {
    fetch.mockReturnValue(null);
    render(<ChartConfigModal {...props} />);

    await waitFor(() => {
      expect(document.querySelector('.MuiPaper-root')).toBeNull();
    });
  });

  it('should display message of no data source preset if api return empty datasource array', async () => {
    fetch.mockReturnValue({ dataSources: [] });
    render(<ChartConfigModal {...props} />);

    await waitFor(() => {
      expect(document.querySelector('.MuiPaper-root')).toHaveTextContent(
        /No data source present, upload data source/i,
      );
    });
  });
});
