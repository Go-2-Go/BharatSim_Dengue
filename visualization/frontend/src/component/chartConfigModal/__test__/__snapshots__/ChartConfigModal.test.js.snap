// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartConfigModal /> should match a snapshot for <ChartConfigModal /> 1`] = `
<div
  aria-labelledby="form-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root"
    id="form-dialog-title"
  >
    <h2
      class="MuiTypography-root MuiTypography-h6"
    >
      Chart Config
    </h2>
  </div>
  <div
    class="MuiDialogContent-root"
  >
    <div
      class="MuiBox-root MuiBox-root-2 ChartConfigModal-root-1"
    >
      <div
        class="MuiFormControl-root makeStyles-formControl-3"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          id="dropdown-label"
        >
          select data source
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          data-testid="dropdown-dataSources"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="dropdown-label dropdown-dataSources"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            id="dropdown-dataSources"
            role="button"
            tabindex="0"
          >
            <span>
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                select data source
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
      data-testid="button-ok"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Ok
      </span>
    </button>
  </div>
</div>
`;
